executor.go

```diff
@@ -48,9 +51,160 @@ type ValidationResult struct {
+	if err != nil {
+		return nil, fmt.Errorf("failed to evaluate verifier reports: %w", err)
+	}
+
+	return &ValidationResult{
+		Succeeded:       decision,
+		ArtifactReports: aggregatedVerifierReports,
+	}, nil
+}
+
+// aggregateVerifierReports generates and aggregates all verifier reports.
+func (e *Executor) aggregateVerifierReports(ctx context.Context, opts ValidateArtifactOptions) ([]*ValidationReport, error) {
+	var aggregatedVerifierReports []*ValidationReport
+	var err error
+
+	// TODO: Implement a worker pool to validate artifacts concurrently.
+	// TODO: Enforce check on the stack size.
+	taskStack := make([]*task, 0)
+
+	// Iterate over all stores to create tasks for each store so that further
+	// referrer artifacts will be fetched from the same store.
+	for _, store := range e.Stores {
+		taskStack = append(taskStack, &task{
+			store:    store,
+			artifact: opts.Subject,
+		})
+	}
+
+	for len(taskStack) > 0 {
+		task := taskStack[len(taskStack)-1]
+		taskStack = taskStack[:len(taskStack)-1]
+
+		taskStack, aggregatedVerifierReports, err = e.processTask(ctx, task, taskStack, aggregatedVerifierReports, opts.ReferenceTypes)
+		if err != nil {
+			return nil, fmt.Errorf("failed to validate artifact %s: %w", task.artifact, err)
+		}
+	}
+
+	return aggregatedVerifierReports, nil
+}
+
+// processTask processes the task and returns the updated stack and aggregated
+// verifier reports.
+func (e *Executor) processTask(ctx context.Context, task *task, stack []*task, aggregatedVerifierReports []*ValidationReport, referenceTypes []string) ([]*task, []*ValidationReport, error) {
+	artifactRef := task.artifact
+	ref, err := registry.ParseReference(artifactRef)
+	if err != nil {
+		return stack, nil, fmt.Errorf("failed to parse artifact reference %s: %w", artifactRef, err)
+	}
+
+	artifactDesc, err := task.store.Resolve(ctx, ref.String())
+	if err != nil {
+		return stack, nil, fmt.Errorf("failed to resolve artifact reference %s: %w", ref.Reference, err)
+	}
+	ref.Reference = artifactDesc.Digest.String()
+
+	stack, artifactReports, err := e.verifyAgainstReferrers(ctx, task.store, ref, stack, referenceTypes)
+	if err != nil {
+		return stack, nil, err
+	}
+
+	// If the current task is the root task, add the artifactReports to the
+	// aggregatedVerifierReports. Otherwise, they are just artifactReports of
+	// the subject artifact.
+	if task.subjectReport != nil {
+		task.subjectReport.ArtifactReports = append(task.subjectReport.ArtifactReports, artifactReports...)
+	} else {
+		aggregatedVerifierReports = append(aggregatedVerifierReports, artifactReports...)
+	}
+	return stack, aggregatedVerifierReports, nil
+}
+
+// verifyAgainstReferrers verifies the subject artifact against all referrers in
+// the store and produces new tasks for each referrer.
+func (e *Executor) verifyAgainstReferrers(ctx context.Context, store Store, subject registry.Reference, stack []*task, referenceTypes []string) ([]*task, []*ValidationReport, error) {
```

Comment:
just a suggestion on the name, referrer provides the artifact, the verifier is responsible for verification, verificationAgainstReferrer may suggest Referrer owns the verification. Shall we keep the previous name , "verifySubject"?

Reply:
stack have strong ties to datastructure , suggestion to rename variable stack to be informative